### PR TITLE
ODE+LSTM

### PR DESCRIPTION
**Pull request by darabos**
I've successfully trained a model with this. The model is awful. 😅 The next steps are:
- Debug until I get an acceptable model.
- Clean up all the issues I swept under the rug, like caching.
- Kill the one-by-one boxes. I don't want them. But `torchdiffeq` can't take different timestamps for different samples in a single tensor. This basically means a batch size of 1. Maybe it's time I implement batching!
- I'm currently testing with an internal dataset. I'll create a public demo. (Or share the internal dataset if possible.)

### CODE REVIEW

File: lynxkite-graph-analytics/src/lynxkite_graph_analytics/ml_ops.py [hunk 1]
```diff
@@ -79,7 +80,46 @@ def train_model(
...
+    """
+    if not input_mapping:
+        return
+    m = bundle.other[model_name].copy()
+    b = bundle
+    inputs = {}
+    for k, v in input_mapping.map.items():
+        if v.df in b.dfs and v.column in b.dfs[v.df]:
+            inputs[k] = b.dfs[v.df][v.column]
+    [num_samples] = set(len(v) for v in inputs.values())
```
[the remark this answers is not in the document]
Comment (darabos) [reply lynxkite-graph-analytics/src/lynxkite_graph_analytics/ml_ops.py:104]: > Using destructuring assignment with a set to obtain num_samples can be fragile if input lengths vary; consider explicitly checking that all input tensors have the same length or extracting the sample count in a more robust manner.

It's a matter of preference, but I like to use `[x] = ...` as a way to say "assert there's one element and take it. When I see `num_samples = lengths[0]` it's unclear that we get correct behavior when `len(lengths) > 1`. You can add an assert or an `if` separately, but what if the code is later changed and those lines are no longer in sync. It's better to have the assertion and the retrieval as a combined atomic step.

**Review by JMLizano — COMMENTED** [review]
Looking good! Nice to see that we arrived to similar ideas, like creating  wrappers to transform operations into `Modules`

**Comment by darabos** [comment]
> * Clean up all the issues I swept under the rug, like caching.

Wow, caching was a trip! I think it's all good now. For future reference, Joblib uses `joblib.hashing.hash(args)` to create the hash key. This will be the name of the directory, like `.joblib-cache/joblib/lynxkite_graph_analytics/ml_ops/model_inference/e01d77a88f40f566db9d500b152734c8/`. `hash()` uses pickling plus some magic for Numpy arrays. It works well for Pandas DataFrames, but it fails for PyTorch tensors. (https://github.com/joblib/joblib/issues/1282) When those are pickled, they save the memory location too, so views can be unpickled correctly. (If I understand correctly.) But `torch.save()` somehow deals with this and produces stable output.

Our models also contain lambdas and random weights, which also mess up the process. So I added `ModelConfig.__getstate__()` which saves the model definition workspace instead of the model. Plus the `torch.save()`-pickled model weights, if the model is already trained.

Even with this, we need to be careful to produce deterministic output, otherwise caching is easily defeated. I had to add a fixed random seed to "Train/test split", and I had to put the model inputs/outputs in alphabetic order.

The cached outputs are also stored in a pickle, so these changes also affected that.

> * Kill the one-by-one boxes. I don't want them. But `torchdiffeq` can't take different timestamps for different samples in a single tensor. This basically means a batch size of 1. Maybe it's time I implement batching!

This is done.

I had to introduce a new "Output sequence" box in the model designer. The reason is that normally we expect the model to output N values for a batch size of N. But with ODE we get T values, which is the number of timestamps, and it's different for each sample. I'm not sure this is a perfect solution — let's adjust it as needed for future models. But I think it makes sense to indicate different output types, like we indicate different input types.

**Comment by darabos** [comment]
I still want to work on this, because it's giving super inaccurate predictions. But this PR has a lot of good changes that I think can go in.

One thing I'm unsure about is the `ode_lstm.py` file. Some of those are quite general. Should I move them to the "standard library" in `lynxkite_ops.py`?

I decided not to do it today mainly because they operate on DataFrames. Ops that take DataFrames as their input are a bit awkward to use. They don't work with a general Bundle input. Maybe we could automatically add a table selector parameter for such boxes? And the DataFrame they return would replace the same DataFrame.
